lib/winpki/ldap.go

```diff
@@ -174,35 +182,144 @@ func convertLDAPError(err error) error {
 	return err
 }
 
-// GetActiveDirectorySID makes an LDAP query to retrieve the security identifier (SID)
-// for the specified Active Directory user.
-func (l *LDAPClient) GetActiveDirectorySID(ctx context.Context, username string) (string, error) {
-	filter := CombineLDAPFilters([]string{
+// GetActiveDirectorySIDAndDN makes an LDAP query to retrieve the security identifier (SID)
+// for the specified Active Directory user. It also returns their distinguished name.
+func (l *LDAPClient) GetActiveDirectorySIDAndDN(ctx context.Context, username string) (string, string, error) {
+	domain := l.cfg.Domain
+	if strings.Contains(username, "@") {
+		parts := strings.SplitN(username, "@", 2)
+		username = parts[0]
+		domain = parts[1]
+	}
+
+	queries := []func() ([]*ldap.Entry, error){
+		func() ([]*ldap.Entry, error) {
+			//User principal name and configured baseDN
+			filter := fmt.Sprintf("(%s=%s)", AttrUserPrincipalName, ldap.EscapeFilter(username+"@"+domain))
+			return l.queryLDAP(ctx, filter, username, DomainDN(l.cfg.Domain))
+		},
+		func() ([]*ldap.Entry, error) {
+			//User principal name and baseDN derived from username
+			filter := fmt.Sprintf("(%s=%s)", AttrUserPrincipalName, ldap.EscapeFilter(username+"@"+domain))
+			return l.queryLDAP(ctx, filter, username, DomainDN(domain))
+		},
+		func() ([]*ldap.Entry, error) {
+			//sAMAccountName and baseDN derived from username
+			if len(username) > 20 {
+				l.cfg.Logger.WarnContext(ctx, "username used for querying sAMAccountName is longer than 20 characters, results might be invalid", "username", username)
+				username = username[:20]
+			}
+			filter := fmt.Sprintf("(%s=%s)", AttrSAMAccountName, ldap.EscapeFilter(username))
+			return l.queryLDAP(ctx, filter, username, DomainDN(domain))
+		},
+	}
+
+	var err error
+	var entries []*ldap.Entry
+	for _, query := range queries {
+		entries, err = query()
+		if err != nil {
+			return "", "", trace.Wrap(err)
+		}
+		if len(entries) > 0 {
+			break
+		}
+		l.cfg.Logger.DebugContext(ctx, "found 0 entries, trying another query")
```

Comment:
Can we add some context to this message that shows _what_ we were trying to query that resulted in 0 entries?

Reply:
Added query name/description to the log message